core/templates/dev/head/exploration_editor/EditorServices.js

```diff
@@ -1952,3 +2012,294 @@ oppia.factory('explorationWarningsService', [
+        }
+      } else if (!utilsService.isEmpty(newValue)) {
+        result = 'added';
+      }
+    }
+    return result;
+  };
+
+  var makeHumanReadable = function(lostChanges) {
+    var outerHtml = angular.element('<ul></ul>');
+    var stateWiseEditsMapping = {};
+    // The variable stateWiseEditsMapping stores the edits grouped by state.
+    // For instance, you made the following edits:
+    // 1. Changed content to 'Welcome!' instead of '' in 'First Card'.
+    // 2. Added an interaction in this state.
+    // 2. Added a new state 'End'.
+    // 3. Ended Exporation from state 'End'.
+    // stateWiseEditsMapping will look something like this:
+    // - 'First Card': [
+    //   - 'Edited Content: Welcome!',:
+    //   - 'Added Interaction: Continue',
+    //   - 'Added interaction customizations']
+    // - 'End': ['Ended exploration']
+
+    lostChanges.forEach(function(lostChange) {
+      switch (lostChange.cmd) {
+        case CMD_ADD_STATE:
+          outerHtml.append(
+            angular.element('<li></li>').html(
+              'Added state: ' + lostChange.state_name));
+          break;
+        case CMD_RENAME_STATE:
+          outerHtml.append(
+            angular.element('<li></li>').html(
+              'Renamed state: ' + lostChange.old_state_name + ' to ' +
+                lostChange.new_state_name));
+          break;
+        case CMD_DELETE_STATE:
+          outerHtml.append(
+            angular.element('<li></li>').html(
+              'Deleted state: ' + lostChange.state_name));
+          break;
+        case CMD_EDIT_STATE_PROPERTY:
+          var newValue = getStatePropertyValue(lostChange.new_value);
+          var oldValue = getStatePropertyValue(lostChange.old_value);
+          var stateName = lostChange.state_name;
+          if (!stateWiseEditsMapping[stateName]) {
+            stateWiseEditsMapping[stateName] = [];
+          }
+
+          switch (lostChange.property_name) {
+            case 'content':
+              if (newValue !== null) {
+                stateWiseEditsMapping[stateName].push(
+                  angular.element('<div></div>').html(
+                    'Edited content: ' + newValue.value));
+              }
+              break;
+
+            case 'widget_id':
+              if (oldValue === null) {
+                if (newValue !== 'EndExploration') {
+                  lostChangeValue = 'Added Interaction: ' + newValue;
+                } else {
+                  lostChangeValue = 'Ended Exploration';
+                }
+              } else {
+                lostChangeValue = 'Deleted Interaction: ' + oldValue;
+              }
+              stateWiseEditsMapping[stateName].push(
+                angular.element('<div></div>').html(lostChangeValue));
+              break;
+
+            case 'widget_customization_args':
+              if (utilsService.isEmpty(oldValue)) {
+                lostChangeValue = 'Added Interaction Customizations';
```

Comment:
I think these will all need to start with var now, given the variable is no longer created at the start.